org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java

```diff
@@ -280,6 +285,143 @@ public void EntityDeclUnterminated() throws Exception {
+				"\t<!ENTITY nbsp \"entity-value\">\r\n" +
+				"]>\r\n")));
+	}
+
+	@Test
+	public void EntityNotDeclaredDoctypeNoSubset() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article >\n" + //
+				"<article>\n" + //
+				"	&nbsp;\n" + //
+				"</article>";
+
+		Diagnostic d = d(3, 1, 3, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+
+		testCodeActionsFor(xml, d,
+				ca(d, te(1, 18, 1, 18, "[\n\t<!ENTITY nbsp \"entity-value\">\n]")));
+	}
+
+	@Test
+	public void EntityNotDeclaredDoctypeNoSubsetNoSpace() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article>\n" + //
+				"<article>\n" + //
+				"	&nbsp;\n" + //
+				"</article>";
+
+		Diagnostic d = d(3, 1, 3, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+
+		testCodeActionsFor(xml, d,
+				ca(d, te(1, 17, 1, 17, " [\n\t<!ENTITY nbsp \"entity-value\">\n]")));
+	}
+
+	@Test
+	public void EntityNotDeclaredDoctypeEmptySubset() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article []>\n" + //
+				"<article>\n" + //
+				"	&nbsp;\n" + //
+				"</article>";
+
+		Diagnostic d = d(3, 1, 3, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+
+		testCodeActionsFor(xml, d,
+				ca(d, te(1, 19, 1, 19, "\n\t<!ENTITY nbsp \"entity-value\">\n")));
+	}
+
+	@Test
+	public void EntityNotDeclaredDoctypeEmptySubsetWithNewline() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article [\n" + //
+				"]>\n" + //
+				"<article>\n" + //
+				"	&nbsp;\n" + //
+				"</article>";
+
+		Diagnostic d = d(4, 1, 4, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+
+		testCodeActionsFor(xml, d,
+				ca(d, te(2, 0, 2, 0, "\t<!ENTITY nbsp \"entity-value\">\n")));
+	}
+
+	@Test
+	public void EntityNotDeclaredSingleQuotes() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" + //
+				"<!DOCTYPE article [\r\n" + //
+				"	<!ELEMENT article (#PCDATA)>\r\n" + //
+				"]>\r\n" + //
+				"<article>\r\n" + //
+				"	&nbsp;\r\n" + //
+				"</article>";
+		SharedSettings settings = new SharedSettings();
+		settings.getPreferences().setQuoteStyle(QuoteStyle.singleQuotes);
+		settings.getFormattingSettings().setInsertSpaces(false);
+		Diagnostic d = d(5, 1, 5, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+		testCodeActionsFor(xml, d, settings, ca(d, te(2, 29, 2, 29, "\r\n\t<!ENTITY nbsp \'entity-value\'>")));
+	}
+

```

Comment:
```xml
<?xml version="1.0" encoding="UTF-8"?><text>
<text>
  &c;
</text>
</text>
```
The code action will put the DOCTYPE declaration after the first `<text>` tag, on the second line:

![make-DTD-entity-code-action](https://user-images.githubusercontent.com/22376627/83052570-82f7e000-a01d-11ea-8661-d18ad826f852.gif)

If it makes sense to fix this issue, then we should test for it here.

Reply:
Thanks, fixed